tests/functional/suite_test.go

```diff
@@ -185,15 +183,9 @@ var _ = BeforeSuite(func() {
 	}()
 
 	// wait for the webhook server to get ready
-	dialer := &net.Dialer{Timeout: time.Duration(10) * time.Second}
-	addrPort := fmt.Sprintf("%s:%d", webhookInstallOptions.LocalServingHost, webhookInstallOptions.LocalServingPort)
 	Eventually(func() error {
-		conn, err := tls.DialWithDialer(dialer, "tcp", addrPort, &tls.Config{InsecureSkipVerify: true})
-		if err != nil {
-			return err
-		}
-		conn.Close()
-		return nil
+		checker := k8sManager.GetWebhookServer().StartedChecker()
```

Comment:
I'm not sure this actually does what we expect. I tried to comment out `err = (&placementv1.PlacementAPI{}).SetupWebhookWithManager(k8sManager)` above to see if this new check catches the missing webhook but it does not.

Reply:
Webhooks should be installed to use this check. This check internally executes the same connection check we currently have. See https://github.com/kubernetes-sigs/controller-runtime/blob/main/pkg/webhook/server.go#L308-L327 .

Reply:
Something is still different between the two solution. 

1. If on main I disable the webhook:
```
diff --git a/tests/functional/suite_test.go b/tests/functional/suite_test.go
index 750b272..05e5e3a 100644
--- a/tests/functional/suite_test.go
+++ b/tests/functional/suite_test.go
@@ -165,8 +165,8 @@ var _ = BeforeSuite(func() {
        kclient, err := kubernetes.NewForConfig(cfg)
        Expect(err).ToNot(HaveOccurred(), "failed to create kclient")
 
-       err = (&placementv1.PlacementAPI{}).SetupWebhookWithManager(k8sManager)
-       Expect(err).NotTo(HaveOccurred())
+       //      err = (&placementv1.PlacementAPI{}).SetupWebhookWithManager(k8sManager)
+       //      Expect(err).NotTo(HaveOccurred())
 
        placementv1.SetupDefaults()
```
and then run a test case the test setup fails at the dialer:
```
$ make test GINKGO_ARGS="-v --procs=1 --focus='A PlacementAPI instance is created without container images'"
[...snip...]
Will run 1 of 9 specs
------------------------------
[BeforeSuite] 
/home/gibi/upstream/git/openstack-k8s-operators/placement-operator/tests/functional/suite_test.go:84
  STEP: bootstrapping test environment @ 07/14/23 10:57:36.494
  2023-07-14T10:57:36+02:00	DEBUG	controller-runtime.test-env	starting control plane
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	reading Webhooks from path	{"path": "../../config/webhook"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	read webhooks from file	{"file": "kustomization.yaml"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	read webhooks from file	{"file": "kustomizeconfig.yaml"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	read webhooks from file	{"file": "manifests.yaml"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	read webhooks from file	{"file": "service.yaml"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	installing CRDs
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	reading CRDs from path	{"path": "../../config/crd/bases"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	read CRDs from file	{"file": "placement.openstack.org_placementapis.yaml"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	reading CRDs from path	{"path": "/home/gibi/go/pkg/mod/github.com/openstack-k8s-operators/keystone-operator/api@v0.0.0-20230622141005-e9220a4b3dfe/bases"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	read CRDs from file	{"file": "keystone.openstack.org_keystoneapis.yaml"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	read CRDs from file	{"file": "keystone.openstack.org_keystoneendpoints.yaml"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	read CRDs from file	{"file": "keystone.openstack.org_keystoneservices.yaml"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	reading CRDs from path	{"path": "/home/gibi/go/pkg/mod/github.com/openstack-k8s-operators/mariadb-operator/api@v0.0.0-20230622153114-756aead1d819/bases"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	read CRDs from file	{"file": "mariadb.openstack.org_galeras.yaml"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	read CRDs from file	{"file": "mariadb.openstack.org_mariadbdatabases.yaml"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	read CRDs from file	{"file": "mariadb.openstack.org_mariadbs.yaml"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	reading CRDs from path	{"path": "/home/gibi/go/pkg/mod/github.com/openstack-k8s-operators/lib-common/modules/test@v0.0.0-20230711124224-54bd95cadcf3/openshift_crds/route/v1"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	read CRDs from file	{"file": "route_crd.yaml"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	installing CRD	{"crd": "galeras.mariadb.openstack.org"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	installing CRD	{"crd": "mariadbdatabases.mariadb.openstack.org"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	installing CRD	{"crd": "mariadbs.mariadb.openstack.org"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	installing CRD	{"crd": "routes.route.openshift.io"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	installing CRD	{"crd": "placementapis.placement.openstack.org"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	installing CRD	{"crd": "keystoneapis.keystone.openstack.org"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	installing CRD	{"crd": "keystoneendpoints.keystone.openstack.org"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	installing CRD	{"crd": "keystoneservices.keystone.openstack.org"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	adding API in waitlist	{"GV": "mariadb.openstack.org/v1beta1"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	adding API in waitlist	{"GV": "mariadb.openstack.org/v1beta1"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	adding API in waitlist	{"GV": "mariadb.openstack.org/v1beta1"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	adding API in waitlist	{"GV": "route.openshift.io/v1"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	adding API in waitlist	{"GV": "placement.openstack.org/v1beta1"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	adding API in waitlist	{"GV": "keystone.openstack.org/v1beta1"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	adding API in waitlist	{"GV": "keystone.openstack.org/v1beta1"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	adding API in waitlist	{"GV": "keystone.openstack.org/v1beta1"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	installing webhooks
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	installing mutating webhook	{"webhook": "mutating-webhook-configuration"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	installing validating webhook	{"webhook": "validating-webhook-configuration"}
  2023-07-14T10:57:39+02:00	INFO	placementapi-resource	PlacementAPI defaults initialized	{"defaults": {"ContainerImageURL":"quay.io/podified-antelope-centos9/openstack-placement-api:current-podified"}}
  2023-07-14T10:57:39+02:00	INFO	Starting EventSource	{"controller": "placementapi", "controllerGroup": "placement.openstack.org", "controllerKind": "PlacementAPI", "source": "kind source: *v1beta1.PlacementAPI"}
  2023-07-14T10:57:39+02:00	INFO	Starting EventSource	{"controller": "placementapi", "controllerGroup": "placement.openstack.org", "controllerKind": "PlacementAPI", "source": "kind source: *v1beta1.MariaDBDatabase"}
  2023-07-14T10:57:39+02:00	INFO	Starting EventSource	{"controller": "placementapi", "controllerGroup": "placement.openstack.org", "controllerKind": "PlacementAPI", "source": "kind source: *v1beta1.KeystoneService"}
  2023-07-14T10:57:39+02:00	INFO	Starting EventSource	{"controller": "placementapi", "controllerGroup": "placement.openstack.org", "controllerKind": "PlacementAPI", "source": "kind source: *v1beta1.KeystoneEndpoint"}
  2023-07-14T10:57:39+02:00	INFO	Starting EventSource	{"controller": "placementapi", "controllerGroup": "placement.openstack.org", "controllerKind": "PlacementAPI", "source": "kind source: *v1.Job"}
  2023-07-14T10:57:39+02:00	INFO	Starting EventSource	{"controller": "placementapi", "controllerGroup": "placement.openstack.org", "controllerKind": "PlacementAPI", "source": "kind source: *v1.Service"}
  2023-07-14T10:57:39+02:00	INFO	Starting EventSource	{"controller": "placementapi", "controllerGroup": "placement.openstack.org", "controllerKind": "PlacementAPI", "source": "kind source: *v1.Secret"}
  2023-07-14T10:57:39+02:00	INFO	Starting EventSource	{"controller": "placementapi", "controllerGroup": "placement.openstack.org", "controllerKind": "PlacementAPI", "source": "kind source: *v1.ConfigMap"}
  2023-07-14T10:57:39+02:00	INFO	Starting EventSource	{"controller": "placementapi", "controllerGroup": "placement.openstack.org", "controllerKind": "PlacementAPI", "source": "kind source: *v1.Deployment"}
  2023-07-14T10:57:39+02:00	INFO	Starting EventSource	{"controller": "placementapi", "controllerGroup": "placement.openstack.org", "controllerKind": "PlacementAPI", "source": "kind source: *v1.Route"}
  2023-07-14T10:57:39+02:00	INFO	Starting EventSource	{"controller": "placementapi", "controllerGroup": "placement.openstack.org", "controllerKind": "PlacementAPI", "source": "kind source: *v1.ServiceAccount"}
  2023-07-14T10:57:39+02:00	INFO	Starting EventSource	{"controller": "placementapi", "controllerGroup": "placement.openstack.org", "controllerKind": "PlacementAPI", "source": "kind source: *v1.Role"}
  2023-07-14T10:57:39+02:00	INFO	Starting EventSource	{"controller": "placementapi", "controllerGroup": "placement.openstack.org", "controllerKind": "PlacementAPI", "source": "kind source: *v1.RoleBinding"}
  2023-07-14T10:57:39+02:00	INFO	Starting Controller	{"controller": "placementapi", "controllerGroup": "placement.openstack.org", "controllerKind": "PlacementAPI"}
  2023-07-14T10:57:39+02:00	INFO	Starting workers	{"controller": "placementapi", "controllerGroup": "placement.openstack.org", "controllerKind": "PlacementAPI", "worker count": 1}
  [FAILED] in [BeforeSuite] - /home/gibi/upstream/git/openstack-k8s-operators/placement-operator/tests/functional/suite_test.go:197 @ 07/14/23 10:57:40.646
[BeforeSuite] [FAILED] [4.153 seconds]
[BeforeSuite] 
/home/gibi/upstream/git/openstack-k8s-operators/placement-operator/tests/functional/suite_test.go:84

  [FAILED] Timed out after 1.001s.
  Expected success, but got an error:
      <*net.OpError | 0xc000559ef0>: 
      dial tcp 127.0.0.1:40081: connect: connection refused
      {
          Op: "dial",
          Net: "tcp",
          Source: nil,
          Addr: <*net.TCPAddr | 0xc000930960>{
              IP: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 255, 255, 127, 0, 0, 1],
              Port: 40081,
              Zone: "",
          },
          Err: <*os.SyscallError | 0xc0001e50c0>{
              Syscall: "connect",
              Err: <syscall.Errno>0x6f,
          },
      }
  In [BeforeSuite] at: /home/gibi/upstream/git/openstack-k8s-operators/placement-operator/tests/functional/suite_test.go:197 @ 07/14/23 10:57:40.646

  Full Stack Trace
    github.com/openstack-k8s-operators/placement-operator/tests/functional_test.glob..func3()
    	/home/gibi/upstream/git/openstack-k8s-operators/placement-operator/tests/functional/suite_test.go:197 +0x13c7
------------------------------
[AfterSuite] 
/home/gibi/upstream/git/openstack-k8s-operators/placement-operator/tests/functional/suite_test.go:200
  STEP: tearing down the test environment @ 07/14/23 10:57:40.647
  2023-07-14T10:57:40+02:00	INFO	Stopping and waiting for non leader election runnables
  2023-07-14T10:57:40+02:00	INFO	Stopping and waiting for leader election runnables
  2023-07-14T10:57:40+02:00	INFO	Shutdown signal received, waiting for all workers to finish	{"controller": "placementapi", "controllerGroup": "placement.openstack.org", "controllerKind": "PlacementAPI"}
  2023-07-14T10:57:40+02:00	INFO	All workers finished	{"controller": "placementapi", "controllerGroup": "placement.openstack.org", "controllerKind": "PlacementAPI"}
  2023-07-14T10:57:40+02:00	INFO	Stopping and waiting for caches
  2023-07-14T10:57:40+02:00	INFO	Stopping and waiting for webhooks
  2023-07-14T10:57:40+02:00	INFO	Wait completed, proceeding to shutdown the manager
[AfterSuite] PASSED [1.124 seconds]
------------------------------
[...snip...]
```
so the current check catches a missing webhook.

2. Then if I apply the same diff to the the current PR then the new check passes and only the actual test case fails later when try to call the webhook. 

This shows me that the current check is doing more than the proposed one.

Reply:
I think this behavior is caused by `GetWebhookServer` which directs the request to controller manager.
 https://github.com/kubernetes-sigs/controller-runtime/blob/release-0.14/pkg/manager/internal.go#L309-L316

This behavior was changed in 0.15 but we are still using old 0.14 for now.
 https://github.com/kubernetes-sigs/controller-runtime/commit/0377e0f7a41bd9d88d43811192309b193c9cd793

We can probably add some checks to ensure we are sending the requests to the right server, but I'm wondering if it's worth effort here, because
* we are catching the failure even in an ugly way
* this will be improved in the new controller-runtime version

@gibizer May I know your thoughts on this ?

Reply:
Can we wait for the controller-runtime@0.15 bump before we change this? Or the change in main.go fixes an immediate problem with our operators?

Reply:
I can remove this change now. I thought using the built-in function can simplify the implementation but that's not as not urgent as the original problem.